Vectorized complex unary and binary op support.

dylanbespalko opened the pull request:
Added Complex support with AVX to unary ops and binary ops.

I need to add nan propagation to minimum() and maximum() in the future.
In-tree changes to pytorch to support complex numbers are being submitted here.
Out-of-tree support for complex numbers is here: pytorch-cpu-strided-complex extension

Preliminary Benchmarks are here.

I tried rrii and riri and found that riri is better in most situations.
Divide is very slow because you can't reduce 1/(x+y)
Sqrt is also very slow.
Reciprocal could be sped up after I add conj()
Everything else is typically within 20% of the real number performance.
Questions:

Why does macOS not support mil? #if AT_MKL_ENABLED() && !defined(__APPLE__) in vml.h. MKL does support some complex operations like Abs, so I was curious about trying it.
Is MKL just calling AVX?

dylanbespalko committed:
Modified existing Vec256 classes to support new operations for complex numbers

dylanbespalko committed:
Modified existing Vec256 classes to support new operations for complex numbers2

dylanbespalko committed:
Added new unary ops for complex number support

dylanbespalko committed:
Added complex double Vec256 support

dylanbespalko committed:
Minor fixes to complex riri format

dylanbespalko committed:
Fixing CI Failure with VML.h

dylanbespalko committed:
Added complex_float_riri

dylanbespalko committed:
Added vec256_complex_float

dylanbespalko committed:
Resolved merge conflicts that occured when lgamma was added to UnaryOps

dylanbespalko committed:
Merged changes from pytorch master

dylanbespalko commented:
@ezyang, @VitalyFedyunin

Added Complex support with AVX to unary ops and binary ops.
- I need to add nan propagation to `minimum()` and `maximum()` in the future.

In-tree changes to pytorch to support complex numbers are being submitted here.
Out-of-tree support for complex numbers is here: [pytorch-cpu-strided-complex extension](https://gitlab.com/pytorch-complex/pytorch-cpu-strided-complex)

Preliminary Benchmarks are [here](https://docs.google.com/spreadsheets/d/17pObcrSTpV4BOOX9FYf1vIX3QUlEgQhLvL1IBEyJyzs/edit?usp=sharing).
- I tried `rrii` and `riri` and found that `riri` is better in most situations.
- Divide is very slow because you can't reduce `1/(x+y)`
- Sqrt is also very slow.
- Reciprocal could be sped up after I add `conj()`
- Everything else is typically within 20% of the real number performance.

Questions:
1. Why does macOS not support mil? `#if AT_MKL_ENABLED() && !defined(__APPLE__) in vml.h`. MKL does support some complex operations like Abs, so I was curious about trying it.
2. Is MKL just calling AVX?

ezyang commented on aten/src/ATen/cpu/vec256/vec256_base.h:
```diff
@@ -390,24 +407,30 @@ inline T minimum(const T& a, const T& b) {
+  value_t (*zabs_)(T) = at::native::zabs;
   for (int i = 0; i != Vec256<T>::size(); i++) {
-    c[i] = a[i] < min_vec[i] ? min_vec[i] : (a[i] > max_vec[i] ? max_vec[i] : a[i]);
+    c[i] = zabs_(a[i]) < zabs_(min_vec[i]) ? min_vec[i] : (zabs_(a[i]) > zabs_(max_vec[i]) ? max_vec[i] : a[i]);
```
Note for reviewers: `zabs` is no op on real only numbers. (The semantics of `zabs` is pretty weird. In particular, `zabs(-1) == -1` if `-1` is a non-complex type, but `== 1` if it is a complex type. That seems very fishy, semantically to me. I'd prefer not to implement clamp on complex unless we have a really good reason for hacking it up this way.

dylanbespalko replied:
Hmm, I was under the impression that the min, max arguments could be tensors, but the torch docs indicate that they are just numeric constants.

The following is a common voltage control mechanism that is commonly used in my work:
`np.where(np.where(np.abs(a) > np.abs(min), a, min) < np.abs(max), max)`

Perhaps I should just use `torch.where()`.

ezyang commented on aten/src/ATen/cpu/vec256/vec256_complex_double.h:
```diff
@@ -0,0 +1,342 @@
+    double real_value = std::real(val);
+    double imag_value = std::imag(val);
+    values = _mm256_setr_pd(real_value, imag_value,
+                            real_value, imag_value);
```
I'm going to trust that you got the SIMD instructions correct here. If there is some sort of unit test you could write here that would be spiffy.

dylanbespalko replied:
Yes, the unit tests for complex numbers can be found at [pytorch-cpu-strided-complex extension](https://gitlab.com/pytorch-complex/pytorch-cpu-strided-complex).  Once things are running smoothly I can move them into the pytorch unit tests.

ezyang commented on aten/src/ATen/native/cpu/Loops.h:
```diff
@@ -42,13 +42,13 @@ namespace at { namespace native { namespace {
 using namespace vec256;
 
-template <typename traits, std::size_t... I>
+template <typename traits, std::size_t... INDEX>
```
No substantive changes here, I guess?

dylanbespalko replied:
Yes, this was causing a bunch of CI failures. On some machines `std::complex` #defines the letter `I`. If you include `complex.h` you will get build errors on some machines.

ezyang commented:
Looks good. My comments are just minor. Did not check the vectorization carefully.

> Why does macOS not support mil? #if AT_MKL_ENABLED() && !defined(__APPLE__) in vml.h. MKL does support some complex operations like Abs, so I was curious about trying it.

Condition was added in #8488. @cpuhrsch do you remember why you Xed out apple?

> Is MKL just calling AVX?

dylanbespalko commented on aten/src/ATen/cpu/vec256/vec256_complex_double.h:
```diff
@@ -0,0 +1,342 @@
+    return _mm256_permute_pd(imag_(), 0x05);        //b        a
+  }
+  Vec256<std::complex<double>> acos() const {
+    return map(std::acos);
```
Yeah, I actually couldn't figure that out. I read that template classes don't support inheritance? Am I crazy?

ezyang replied:
Non-virtual template methods inherit just fine. https://godbolt.org/z/Gt1QfK

dylanbespalko replied:
I looked into it. Vec256 is a template class. While inheritance is perfectly valid 
eg `class Vec256z : public Vec256<std:complex<double>>{}` it requires that I define a new class `Vec256z`. The rest of the code is creating template objects using `Vec256<scalar_t>`.  In your example code you needed to create a new class name `B`, whereas I have to use `Vec256`.

You may notice that `Vec256<int64_t>` only defines several methods of `Vec256<T>`. I think this is because very few math kernels use integer specific function calls.

cpuhrsch commented:
Yes, I disabled it for OSX because there was a hard error for mkl VML abs. Instead of fixing this issue at that time I commented it out due to prioritization.

dylanbespalko commented:
@cpuhrsch

>Yes, I disabled it for OSX because there was a hard error for mkl VML abs. Instead of fixing this issue at that time I commented it out due to prioritization.

That should be fine, I deploy on linux anyways. I benchmarked two different memory layouts for complex number support. The secondary solution was better for norm related calculations, however I also know that MKL has complex support for `Abs()`. I was just curious to see if this could replace my primary implementation.

dylanbespalko committed:
merged changes with new UnaryOp that was added

dylanbespalko commented:
@VitalyFedyunin,

Have you had a chance to look at `aten/src/ATen/cpu/vec256/vec256_complex_double.h`?

I'm currently looking into ways to speed up reciprocal and div. You can find my other AVX implementations [here](https://gitlab.com/pytorch-complex/pytorch-cpu-strided-complex/tree/master/vec256).
The AVX implantation used here is: `vec256_complex_double_riri.h`.

VitalyFedyunin commented:
Hi! Sorry I have 7 more PRs to review before this one. I promise to take a look by the end of this week.

VitalyFedyunin commented on aten/src/ATen/cpu/vec256/vec256_complex_double.h:
```diff
@@ -0,0 +1,342 @@
+  }
+  __m256d real_() const {
+    auto mask = _mm256_setr_pd(1.0, 0.0, 1.0, 0.0);
+    return _mm256_mul_pd(values, mask);
```
Bitmask might be faster

dylanbespalko replied:
Excellent suggestion. Done.

VitalyFedyunin reviewed (commented):
I really want to see benchmarking numbers, especially for existing altered ops like `clamp`

dylanbespalko commented:
@VitalyFedyunin 
> I really want to see benchmarking numbers, especially for existing altered ops like `clamp`

Understood. I can provide those later today.

I also found a way to speed up the `div` and `reciprocal` for complex numbers. Those two ops are really bad in this PR.

I'll do some benchmarking and update this branch.

dylanbespalko committed:
merging changes from master and added is_complex_t to add complex support to the Vec256_base abs() method

ezyang commented:
Holler when you want another look

dylanbespalko committed:
Speed improvements to complex reciprocal and div. Angle is now numerically stable between 0-360 deg

dylanbespalko committed:
Added the complex conj() operator

dylanbespalko committed:
Merge branch 'master' of github.com:pytorch/pytorch into unary_binary_ops

dylanbespalko committed:
merged changes from master

dylanbespalko commented:
> I really want to see benchmarking numbers, especially for existing altered ops like `clamp`

Here are the benchmarking numbers for dtype=float64

Build:  BUILD_TEST=0 USE_NATIVE_ARCH=ON python setup.py develop
python -m pt.complex_clamp_test --omp_num_threads 1 --mkl_num_threads 1

----------------------------------------
PyTorch/Caffe2 Operator Micro-benchmarks with zabs()
----------------------------------------
Tag : short

Benchmarking PyTorch: clamp
Mode: Eager
Name: clamp_M512_N512
Input: M: 512, N: 512
Forward Execution Time (us) : 120.053
Forward Execution Time (us) : 119.689
Forward Execution Time (us) : 119.227


 ----------------------------------------
PyTorch/Caffe2 Operator Micro-benchmarks without zabs()
----------------------------------------
Tag : short

Benchmarking PyTorch: clamp
Mode: Eager
Name: clamp_M512_N512
Input: M: 512, N: 512
Forward Execution Time (us) : 119.571
Forward Execution Time (us) : 119.316
Forward Execution Time (us) : 119.235

dylanbespalko committed:
fixes to non-vectorized angle

dylanbespalko commented on aten/src/ATen/native/cpu/UnaryOpsKernel.cpp:
```diff
@@ -29,10 +30,10 @@ namespace {
     cpu_kernel_vec(
         iter,
-        [=](scalar_t a) -> scalar_t { return (1 / (1 + std::exp((-a)))); },
+        [=](scalar_t a) -> scalar_t { return (decltype(a)(1) / (decltype(a)(1) + std::exp((-a)))); },
```
Yes, fixed.

dylanbespalko commented:
@ezyang,

Assuming all goes well with the CI tonight, I think I have made the requested changes. Note I have also modified:

- Modified Vec256::angle() to provide a continuous 0-360 degrees range.
- Modified Vec256::clamp()/minimum()/maximum()/abs() to use enable_if<TYPE> method overloads that don't require ztype() or zabs()
- Modified Vec256::reciprocal()/div() to improve performance.
- Added Vec256::conj
- Modified Vec256 methods to use AND instead of MUL operations to improve performance.

Let me know if you see any problems.

ezyang commented:
The errors in 

```

Oct 04 10:05:45 ======================================================================
Oct 04 10:05:45 ERROR: test_DoubleTensor_lgamma (__main__.TestCuda)
Oct 04 10:05:45 ----------------------------------------------------------------------
Oct 04 10:05:45 Traceback (most recent call last):
Oct 04 10:05:45   File "/var/lib/jenkins/workspace/test/common_utils.py", line 543, in wrapper
Oct 04 10:05:45     method(*args, **kwargs)
Oct 04 10:05:45   File "/var/lib/jenkins/workspace/test/common_utils.py", line 543, in wrapper
Oct 04 10:05:45     method(*args, **kwargs)
Oct 04 10:05:45   File "test_cuda.py", line 637, in tmp
Oct 04 10:05:45     gpu_result = getattr(gpu_tensor, fn)(*gpu_args)
Oct 04 10:05:45 RuntimeError: Expected tensor to have cpu DeviceType, but got tensor with cuda DeviceType (while checking arguments for lgamma)
```

look real-ish.

ezyang commented on aten/src/ATen/cpu/vec256/vec256_base.h:
```diff
@@ -185,6 +193,30 @@ struct Vec256 {
+    return ret;
+  }
+  Vec256<T> angle() const {
+    return *this;
```
These default implementations look wrong lol

ezyang commented on aten/src/ATen/native/UnaryOps.cpp:
```diff
@@ -246,7 +246,8 @@ Tensor& mvlgamma_(Tensor& self, int64_t p) {
   }                                                                    \
   Tensor& _##op##_out_##prefix(Tensor& result, const Tensor& self) {   \
-    checkBackend(#op, result, Backend::device);                        \
+    checkDeviceType(#op, result, DeviceType::CPU);                    \
```
Error probably caused by this change

ezyang commented on benchmarks/operator_benchmark/benchmark_core.py:
```diff
@@ -12,6 +12,7 @@
 # needs to be imported after torch
 import cpp_extension # noqa
 
+import cpp_extension # noqa
```
?

ezyang reviewed (changes requested):
everything looks good, just need to get ci happy

dylanbespalko committed:
simplified Vec256_base abs(), fixed device check of VEC_OPs

dylanbespalko committed:
same as last

dylanbespalko committed:
fixing Vec256_abs on CI

dylanbespalko committed:
Reverting back to previous solution for abs()

dylanbespalko committed:
trying lambda function for abs()

dylanbespalko committed:
fixed angle() function

dylanbespalko commented:
@ezyang,

The CI looks happy now :). Have a good weekend.

ezyang committed:
Merge remote-tracking branch 'origin/master' into unary_binary_ops

dylanbespalko committed:
merging from master (fixing conflict with fbgemm submodule

dylanbespalko committed:
Merge branch 'unary_binary_ops' of github.com:dylanbespalko/pytorch into unary_binary_ops

dylanbespalko committed:
fixing accidental fbgemm submodule changes

dylanbespalko commented:
@ezyang,

The fbgemm submodule changes have been removed from the PR. I don't know how they got in there. I pulled from master and I'm running CI again.
